clients/src/main/java/org/apache/kafka/clients/admin/UserScramCredentialUpsertion.java

```diff
@@ -0,0 +1,99 @@
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.kafka.clients.admin;
+
+import java.math.BigInteger;
+import java.nio.charset.StandardCharsets;
+import java.security.SecureRandom;
+import java.util.Objects;
+
+/**
+ * A request to update/insert a SASL/SCRAM credential for a user.
+ *
+ * @see <a href="https://cwiki.apache.org/confluence/display/KAFKA/KIP-554%3A+Add+Broker-side+SCRAM+Config+API">KIP-554: Add Broker-side SCRAM Config API</a>
+ */
+public class UserScramCredentialUpsertion extends UserScramCredentialAlteration {
+    private final ScramCredentialInfo info;
+    private final byte[] salt;
+    private final byte[] password;
+
+    /**
+     * Constructor that generates a random salt
+     *
+     * @param user the user for which the credential is to be updated/inserted
+     * @param credentialInfo the mechanism and iterations to be used
+     * @param password the password
+     */
+    public UserScramCredentialUpsertion(String user, ScramCredentialInfo credentialInfo, String password) {
+        this(user, credentialInfo, password.getBytes(StandardCharsets.UTF_8));
+    }
+
+    /**
+     * Constructor that generates a random salt
+     *
+     * @param user the user for which the credential is to be updated/inserted
+     * @param credentialInfo the mechanism and iterations to be used
+     * @param password the password
+     */
+    public UserScramCredentialUpsertion(String user, ScramCredentialInfo credentialInfo, byte[] password) {
+        this(user, credentialInfo, password, generateRandomSalt());
+    }
+
+    /**
+     * Constructor that accepts an explicit salt
+     *
+     * @param user the user for which the credential is to be updated/inserted
+     * @param credentialInfo the mechanism and iterations to be used
+     * @param password the password
+     * @param salt the salt to be used
+     */
+    public UserScramCredentialUpsertion(String user, ScramCredentialInfo credentialInfo, byte[] password, byte[] salt) {
+        super(Objects.requireNonNull(user));
+        this.info = Objects.requireNonNull(credentialInfo);
+        this.password = Objects.requireNonNull(password);
+        this.salt = Objects.requireNonNull(salt);
+    }
+
+    /**
+     *
+     * @return the mechanism and iterations
+     */
+    public ScramCredentialInfo credentialInfo() {
+        return info;
+    }
+
+    /**
+     *
+     * @return the salt
+     */
+    public byte[] salt() {
+        return salt;
+    }
+
+    /**
+     *
+     * @return the password
+     */
+    public byte[] password() {
+        return password;
+    }
+
+    private static byte[] generateRandomSalt() {
+        return new BigInteger(130, new SecureRandom()).toString(Character.MAX_RADIX).getBytes(StandardCharsets.UTF_8);
```

Comment:
I think we can avoid a trip through BigInteger with something like this:
```
      SecureRandom random = new SecureRandom();
      byte bytes[] = new byte[length];
      random.nextBytes(bytes);
```

Reply:
@cmccabe  I think the approach you suggest leaves out how to identify `length` which itself needs to be randomized.  I got the current implementation from `org.apache.kafka.common.security.scram.internals.ScramFormatter`.  I would have invoked `ScramFormatter.secureRandomBytes()` directly, but it is not `static` and I did not want to either instantiate an instance or change methods to static (though the class is internal, so I could have done that). I instead replicated the logic here.  The array length ends up being random with this approach, as do the bytes in the array.  Let me know what you think.  Currently I've left this as-is.

Reply:
The latest commit changes the methods on `org.apache.kafka.common.security.scram.internals.ScramFormatter` to be static, and I now reuse that logic.